### PR TITLE
FW-527. Port CC1200dk

### PR DESCRIPTION
**Pull request by jmmunoz86**

### CODE REVIEW

**Comment by twatteyne** [comment]
Very cool! :+1:

File: bsp/boards/MSP430F5438A/SConscript [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+localEnv = env.Clone()
+
+# scons doesn't let us look to parent directories for source, so the
+# bsp/chips/cc1101/radio.c is off limits from this file. To keep things
+# simple, each SConscript file in bsp/chips/* will return a list of objects
+# which can be appended to the source list. Don't forget to specify a
+# variant_dir, or else the build will occur directly in the chips directory.
+
+cc1101 = localEnv.SConscript(
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/SConscript]: CC1101 or CC1200?

File: bsp/boards/MSP430F5438A/SConscript [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import os
+
+Import('env')
+
+localEnv = env.Clone()
+
+# scons doesn't let us look to parent directories for source, so the
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/SConscript]: please remove this comment, which is very generic

File: bsp/boards/MSP430F5438A/SConscript [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import os
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/SConscript]: The MSP430F5438A is a chip, not a board. Please rename this directory to the name of the board.

File: bsp/boards/MSP430F5438A/board.c [hunk 4]
```diff
@@ -0,0 +1,157 @@
+/**
+\brief WSN430v13b-specific definition of the "board" bsp module.
+
+\author Thomas Watteyne <watteyne@eecs.berkeley.edu>, February 2012.
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board.c]: Please replace the author name by you.

File: bsp/boards/MSP430F5438A/board.c [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+int main(void) {
+   return mote_main();
+}
+
+//=========================== public ==========================================
+
+void board_init() {
+   
+  uint16_t ui16IntState;
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board.c]: please use 4-space indentation across your code.

File: bsp/boards/MSP430F5438A/board.c [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+   bsp_timer_init();
+   radio_init();
+   radiotimer_init();
+   
+   // enable interrupts
+   __bis_SR_register(GIE);
+}
+
+void board_sleep() {
+   __bis_SR_register(GIE+LPM0_bits);             // sleep, but leave ACLK on
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board.c]: Why not use LPM3?

File: bsp/boards/MSP430F5438A/board.c [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+//   }
+//   debugpins_isr_clr();
+//}
+
+
+ISR(USCI_A1){
+    debugpins_isr_set();
+    switch(__even_in_range(UCA1IV,4))
+  {
+  case 0:break;                             // Vector 0 - no interrupt
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board.c]: add newline before `break;` statement

File: bsp/boards/MSP430F5438A/board.c [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+//=========================== private =========================================
+
+//=========================== interrupt handlers ==============================
+
+// DACDMA_VECTOR
+
+// PORT2_VECTOR
+
+//ISR(USART1TX){
+//   debugpins_isr_set();
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board.c]: remove commented-out code

File: bsp/boards/MSP430F5438A/board.c [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        if (radio_isr()==KICK_SCHEDULER) {         // radio:  SFD pin [P1.7]
+         __bic_SR_register_on_exit(CPUOFF);
+      }
+   debugpins_isr_clr();
+}
+
+// WDT_VECTOR
+
+//ISR(COMPARATORA) {
+//   debugpins_isr_set();
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board.c]: remove commented-out code

File: bsp/boards/MSP430F5438A/board_info.h [hunk 10]
```diff
@@ -0,0 +1,105 @@
+/**
+\brief WSN430v13b-specific board information bsp module.
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board_info.h]: wrong board name

File: bsp/boards/MSP430F5438A/board.c [hunk 11]
```diff
@@ -0,0 +1,157 @@
+/**
+\brief WSN430v13b-specific definition of the "board" bsp module.
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board.c]: wrong board name. Please apply to all your BSP files.

File: bsp/boards/MSP430F5438A/board_info.h [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+This module file defines board-related element, but which are applicable only
+to this board.
+
+\author Thomas Watteyne <watteyne@eecs.berkeley.edu>, February 2012.
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board_info.h]: wrong author. Please apply throughout your port.

File: bsp/boards/MSP430F5438A/board_info.h [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+//===== adaptive_sync accuracy
+
+#define SYNC_ACCURACY                       1     // ticks
+
+//=========================== variables =======================================
+
+// The variables below are used by CoAP's registration engine.
+
+static const uint8_t rreg_uriquery[]        = "h=ucb";
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board_info.h]: wrong name. UCB is UC Berkeley.

File: bsp/boards/MSP430F5438A/board_info.h [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+static const uint8_t infoBoardname[]        = "WSN430v13b";
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board_info.h]: wrong

File: bsp/boards/MSP430F5438A/bsp.c [hunk 15]
```diff
@@ -0,0 +1,809 @@
+//*****************************************************************************
+//! @file       bsp.c
+//! @brief      Board support package for MSP430F5438a on SmartRF06EB.
+//!
+//! Revised     $Date: 2013-09-18 13:18:48 +0200 (on, 18 sep 2013) $
+//! Revision    $Revision: 10587 $
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/bsp.c]: why do we keep this file in our BSP? Our BSP most probably reimplements all of that, no?

File: bsp/boards/MSP430F5438A/bsp.h [hunk 16]
```diff
@@ -0,0 +1,292 @@
+//*****************************************************************************
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/bsp.h]: same here, why have this file?

File: bsp/boards/MSP430F5438A/bsp_timer.c [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+This functions starts the timer, i.e. the counter increments, but doesn't set
+any compare registers, so no interrupt will fire.
+*/
+void bsp_timer_init() {
+   
+   // clear local variables
+   memset(&bsp_timer_vars,0,sizeof(bsp_timer_vars_t));
+   
+   // set CCRA0 registers
+   //TACCR0               =  0;
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/bsp_timer.c]: remove commented-out unused code

File: bsp/boards/MSP430F5438A/bsp_timer.c [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+   //TACCTL0              =  0;
+   //start TimerA
+  // TA1CCTL0 = CCIE;
+   TA0CTL = TASSEL_1 + MC_2 + TACLR ;  // continuous mode, from ACLK
+
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/bsp_timer.c]: remove unnecessary newlines

File: bsp/boards/MSP430F5438A/debugpins.c [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+//=========================== defines =========================================
+
+//=========================== variables =======================================
+
+//=========================== prototypes ======================================
+
+//=========================== public ==========================================
+
+void debugpins_init() {
+   P6DIR |=  0x20;      // frame [P6.5]
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/debugpins.c]: are those comments accurate? What is the name of that pin on the silkscreen?

File: bsp/boards/MSP430F5438A/eui64.c [hunk 20]
```diff
@@ -0,0 +1,218 @@
+/**
+\brief WSN430v13b-specific definition of the "eui64" bsp module.
+
+\author Thomas Watteyne <watteyne@eecs.berkeley.edu>, March 2012.
+*/
+
+#include "msp430f5438a.h"
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/eui64.c]: Is there a 1-wire EUI64 chip on the board?

File: bsp/boards/MSP430F5438A/leds.c [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+
+//=========================== public ==========================================
+
+void    leds_init() {
+    P4DIR |= BIT0 | BIT1 | BIT2 | BIT3 ;                          // P4DIR = 0bxxxx1111 for LEDs
+    P4OUT |= BIT0 | BIT1 | BIT2 | BIT3 ;                          // P4OUT = 0bxxxx1111, all LEDs off
+   
+}
+
+// red = LED1 = P4.0
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/leds.c]: please confirm that those comments are accurate

File: bsp/boards/MSP430F5438A/radiotimer.c [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+   radiotimer_vars.startFrameCb   = cb;
+}
+
+void radiotimer_setEndFrameCb(radiotimer_capture_cbt cb) {
+   radiotimer_vars.endFrameCb     = cb;
+}
+
+void radiotimer_start(PORT_RADIOTIMER_WIDTH period) {
+   // radio's SFD pin connected to P4.1
+   //P4DIR   &= ~0x02; // input
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/radiotimer.c]: remove code if not used, do NOT comment out

File: bsp/boards/MSP430F5438A/spi.c [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+   memset(&spi_vars,0,sizeof(spi_vars_t));  
+   
+   UCB0CTL1   |=  UCSWRST ;                       // reset state
+   
+   // RF SPI0 CS as GPIO output high
+   //
+   P3SEL &= ~TRXEM_SPI_SC_N_PIN;
+   P3OUT |=  TRXEM_SPI_SC_N_PIN;
+   P3DIR |=  TRXEM_SPI_SC_N_PIN;
+   /* Configuration
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/spi.c]: for multi-line comments, please use `//` rather than `/*`-style comments. This allows us to comment out entire sections with `/*` when debugging.

File: bsp/boards/MSP430F5438A/uart.c [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+
+uart_vars_t uart_vars;
+
+//=========================== prototypes ======================================
+
+//=========================== public ==========================================
+
+void uart_init() {
+   
+   //P3SEL                     =  0x30;            // P3.4,5 = = USCI_A0 TXD/RXD
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/uart.c]: remove

File: bsp/boards/MSP430F5438A/uart.c [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+   P5SEL                       =  0xc0;            //P5.6,7 = USCI_A1
+  
+   UCA1CTL1                  |=  UCSWRST;           // **Put state machine in reset**
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/uart.c]: fix indentation level

File: bsp/boards/MSP430F5438A/uart.c [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+   UCA1CTL1                  |=  UCSSEL_2;          // SMCLK
+   
+   UCA1BR0 = 0xd9;                                 // ~26.5MHz/115200 = 230 = 0xe6
```
Comment (twatteyne) [inline bsp/boards/MSP430F5438A/uart.c]: comment doesn't correspond to code

File: bsp/chips/cc1200/cc1200.c [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+\author Jonathan Munoz <jonathan.munoz@inria.fr>, August 2016.
+*/
+
+#include "board.h"
+#include "CC1200.h"
+#include "spi.h"
+
+extern const cc1200_rf_cfg_t cc1200_rf_cfg;
+
+void CC1200_spiStrobe(uint8_t strobe, cc1200_status_t* statusRead) {
```
Comment (twatteyne) [inline bsp/chips/cc1200/cc1200.c]: by convention, function names start with lowercase characters. Please rename all functions to `cc1200_*`

File: bsp/chips/cc1200/cc1200.c [hunk 28]
```diff
@@ -0,0 +1,228 @@
...
+}
+
+void CC1200_spiWriteReg(uint16_t reg, cc1200_status_t* statusRead, uint8_t regValueToWrite) {
+   
+   uint8_t tempExt  = (uint8_t)(reg>>8);
+   uint8_t tempAddr = (uint8_t)(reg & 0x00FF);
+   
+     /* Decide what register space is accessed */  
+   if(!tempExt)
+   {
```
Comment (twatteyne) [inline bsp/chips/cc1200/cc1200.c]: use format

```
if (foo) {
    poipoi
} else {
    poipoi
}

File: bsp/chips/cc1200/cc1200.c [hunk 29]
```diff
@@ -0,0 +1,228 @@
...
+      SPI_FIRSTBYTE,              // returnType
+      (uint8_t*)statusRead,       // bufRx
+      1,                          // maxLenBufRx
+      SPI_FIRST,                  // isFirst
+      SPI_LAST                    // isLast
+   );
+   }
+   else if (tempExt == 0x2F)
+   {
+   uint8_t              spi_tx_buffer[3];
```
Comment (twatteyne) [inline bsp/chips/cc1200/cc1200.c]: do NOT declare variables in the body of a function, only at the top

File: bsp/chips/cc1200/cc1200.c [hunk 30]
```diff
@@ -0,0 +1,228 @@
...
+      SPI_FIRST,                  // isFirst
+      SPI_LAST                    // isLast
+   );   
+   }
+  
+}
+
+void CC1200_spiReadReg(uint16_t reg, cc1200_status_t* statusRead, uint8_t* regValueRead) {
+  
+   uint8_t tempExt  = (uint8_t)(reg>>8);
```
Comment (twatteyne) [inline bsp/chips/cc1200/cc1200.c]: do not declare and initialize variables in one line.

File: bsp/chips/cc1200/cc1200.c [hunk 31]
```diff
@@ -0,0 +1,228 @@
...
+      pBufRead/*spi_rx_buffer*/,              // bufRx
+      127/*sizeof(spi_rx_buffer)*/,      // maxLenBufRx
+      SPI_FIRST,                  // isFirst
+      SPI_LAST/*SPI_NOTLAST*/                 // isLast
+   );
+   
+ //  *statusRead          = *(cc1200_status_t*)&spi_rx_buffer[0];
+ //  *pLenRead            = spi_rx_buffer[1];
+   
+ //  if (*pLenRead>2 && *pLenRead<=127) {
```
Comment (twatteyne) [inline bsp/chips/cc1200/cc1200.c]: remove unused code

**Comment by twatteyne** [comment]
@jmmunoz86 , I've done a bunch of comments inline in your code. Please go through them and fix them. Some of the comments are style-related and hence apply to all files in your PR, so please propagate the suggestions to all files.

Please make very regular commits (at least a couple per day, not one enormous one per month).

Please remove `.ewd` IAR files, the `.ewp` are enough

File: projects/MSP430F5438A/02drv_openserial/02drv_openserial.ewd [hunk 32]
```diff
@@ -0,0 +1,835 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
Comment (twatteyne) [inline projects/MSP430F5438A/02drv_openserial/02drv_openserial.ewd]: please don't commit `.ewd` files

File: projects/common/01bsp_radio_tx/01bsp_radio_tx.c [hunk 33]
```diff
@@ -21,7 +21,7 @@ remainder of the packet contains an incrementing bytes.
 
 //=========================== defines =========================================
 
-#define LENGTH_PACKET   125+LENGTH_CRC // maximum length is 127 bytes
+#define LENGTH_PACKET   30+LENGTH_CRC // maximum length is 127 bytes
```
Comment (twatteyne) [inline projects/common/01bsp_radio_tx/01bsp_radio_tx.c]: I suggest to use the max length in this little test, as things always break in the corner cases. I understand that the PDU of IEEE802.15.4g is >>127 bytes, but since this application in used on all boards, we should leave it to 127B.

**Comment by twatteyne** [comment]
@jmmunoz86 , you did a fantastic job at taking my comments into account, congratulations!

One thing which is missing at this point is to include this board into the build system, so that one can type:
```
scons board=cc1200dk toolchain=iar oos_openwsn
```
or
```
scons board=cc1200dk toolchain=mspgcc oos_openwsn
```

What else is missing? I just that, I would suggest we merge this PR into the `develop` branch at this point, and work on a different issue. Your thoughts are very welcome.

File: projects/common/01bsp_radio_tx/01bsp_radio_tx.c (hunk 33 above)
Comment (jmmunoz86) [inline projects/common/01bsp_radio_tx/01bsp_radio_tx.c]: yes you're right. I had changed it just to test the TX, which is not yet ready.

**Comment by jmmunoz86** [comment]
thanks @twatteyne ! I still need to fix some of your previous comments and the radio TX. Working on it now.

File: bsp/boards/MSP430F5438A/board_info.h (hunk 13 above)
Comment (jmmunoz86) [inline bsp/boards/MSP430F5438A/board_info.h]: what should it be here?

Comment (twatteyne) [inline bsp/boards/MSP430F5438A/board_info.h]: You can put "TI"

File: projects/common/01bsp_radio_tx/01bsp_radio_tx.c [hunk 34]
```diff
@@ -21,7 +21,7 @@ remainder of the packet contains an incrementing bytes.
 
 //=========================== defines =========================================
 
-#define LENGTH_PACKET   30+LENGTH_CRC // maximum length is 127 bytes
+#define LENGTH_PACKET   124+LENGTH_CRC // maximum length is 127 bytes
```
Comment (twatteyne) [inline projects/common/01bsp_radio_tx/01bsp_radio_tx.c]: that's only 126 bytes,no?

File: bsp/chips/cc1200/cc1200.h [hunk 35]
```diff
@@ -415,52 +410,61 @@ void cc1200_spiReadRam    (uint16_t addr,  cc1200_status_t* statusRead, uint8_t*
...
     {CC1200_SETTLING_CFG,   0x03},
     {CC1200_FS_CFG,         0x12},
     {CC1200_WOR_CFG0,       0x20},
-    {CC1200_WOR_EVENT0_LSB, 0xC3},
-    {CC1200_PKT_CFG2,       0x00},
- //   {CC1200_PKT_CFG2,       0x20},
- //   {CC1200_PKT_CFG1,       0x01},  //added by jmms
+    {CC1200_WOR_EVENT0_LSB, 0x96},
+ //   {CC1200_PKT_CFG2,       0x00},
+    {CC1200_PKT_CFG2,       0x24},  //jmms
```
Comment (twatteyne) [inline bsp/chips/cc1200/cc1200.h]: Please don't sign every line with jmms. IF we all did thins, it would be a mess. You are the author of thta file, so everyone assumes you have either written, or vetted every line in that file.

File: bsp/chips/cc1200/radio.c [hunk 36]
```diff
@@ -138,18 +138,28 @@ void radio_rfOff(void) {
 
 //===== TX
 
-void radio_loadPacket(uint8_t* packet, uint8_t len) {
+void radio_loadPacket(uint8_t* packet, uint16_t len) {
+    //test 802.15.4g PHR. This has to be done by the MAC layer
+    uint8_t  PHR[2];  
+ 
+    if (len<2048){
```
Comment (twatteyne) [inline bsp/chips/cc1200/radio.c]: If's cleaner to do error cheking at the beginning of the function, and return if not OK. Then your normal code is not in an IF statement.

File: bsp/boards/radio.h [hunk 37]
```diff
@@ -66,7 +66,7 @@ void     radio_setFrequency(uint8_t frequency);
 void     radio_rfOn(void);
 void     radio_rfOff(void);
 // TX
-void     radio_loadPacket(uint8_t* packet, uint8_t len);
+void     radio_loadPacket(uint8_t* packet, uint16_t len);
```
Comment (twatteyne) [inline bsp/boards/radio.h]: If you change the signature, you need to change the BSP of all boards. You must have received an e-mail from Travis telling you the builds failed. It's a small change, and I suggest you fix the BSPs of the different boards, until Travis passes again.

File: bsp/chips/cc1200/cc1200.h (hunk 35 above)
Comment (jmmunoz86) [inline bsp/chips/cc1200/cc1200.h]: sorry for having pushed without cleaning the registers I'm testing.

File: bsp/boards/radio.h (hunk 37 above)
Comment (jmmunoz86) [inline bsp/boards/radio.h]: right, the .h file applies to all boards.  On the other hand, I didn't get any e-mail. I'll check the notification settings.

Comment (MrD2) [inline bsp/boards/radio.h]: In fact to be cross platform data types should be size_t for numbers,  uintptr_t for address, void for pointers...

----- Mensagem Original -----
De: "Thomas Watteyne" <notifications@github.com>
Enviado: ‎17/‎08/‎2016 18:16
Para: "openwsn-berkeley/openwsn-fw" <openwsn-fw@noreply.github.com>
Assunto: Re: [openwsn-berkeley/openwsn-fw] FW-527. Port CC1200dk (#280)

In bsp/boards/radio.h:
> @@ -66,7 +66,7 @@ void     radio_setFrequency(uint8_t frequency);
>  void     radio_rfOn(void);
>  void     radio_rfOff(void);
>  // TX
> -void     radio_loadPacket(uint8_t* packet, uint8_t len);
> +void     radio_loadPacket(uint8_t* packet, uint16_t len);
If you change the signature, you need to change the BSP of all boards. You must have received an e-mail from Travis telling you the builds failed. It's a small change, and I suggest you fix the BSPs of the different boards, until Travis passes again.

Comment (twatteyne) [inline bsp/boards/radio.h]: @MrD2, that a good suggestion, but we really want to keep control over the width of the parameters. The change from `uint8_t` to `uint16_t` isn't because that platform uses some different micron-controller, it's because we are now adding support for IEEE802.15.4g, in which link-layer frames can be much longer (max 2047 bytes IIRC) compared to IEEE802.15.4 (127 bytes). So I suggest so still use the explicit `uint16_t` signature

File: bsp/chips/cc1200/cc1200.h (hunk 35 above)
Comment (twatteyne) [inline bsp/chips/cc1200/cc1200.h]: no worries, of course! What I do is that I keep a marker (`poipoi` in my case). I can use it here and there when coding, but know it's temporary. Before committing/pushing, I make sure all the markers are gone. You might do the same.

File: bsp/boards/radio.h (hunk 37 above)
Comment (twatteyne) [inline bsp/boards/radio.h]: @jmmunoz86 , it's strange you didn't get an e-mail. You can look at the bottom of this page, and you'll see a red warning from Travis that at least of the tests failed. If you click on the details, you'll see exactly what failed. The idea is hence for you to keep committing to this fork until the red warning becomes a green checkbox. It's very surprising you didn't get an e-mail. AFAICT, whenever something fails, Travis sends an e-mail to all people who have committed since the last commit on which all tests passed.

Comment (jmmunoz86) [inline bsp/boards/radio.h]: maybe because I had never logged in in travis-ci.org neither with my github credentials nor by using other account.

**Comment by changtengfei** [comment]
@jmmunoz86 , Great job! Congrats on implementing this ports. Currently there is no scons support for the board. I will create an issue for that and merge the code first.
